File: client-java/src/main/java/io/filigran/opencti/entities/Campaign.java
package io.filigran.opencti.entities;

import io.filigran.opencti.OpenCTIApiClient;
import lombok.extern.slf4j.Slf4j;

import java.util.Map;

/**
 * Entity handler for OpenCTI Campaign objects.
 *
 * @author Filigran Team
 * @since 6.9.6
 */
@Slf4j
public class Campaign extends BaseEntity {

    private static final String PROPERTIES = """
            id
            standard_id
            entity_type
            parent_types
            spec_version
            created_at
            updated_at
            createdBy {
                ... on Identity {
                    id
                    standard_id
                    entity_type
                    name
                }
            }
            objectMarking {
                id
                standard_id
                entity_type
                definition_type
                definition
                x_opencti_order
                x_opencti_color
            }
            objectLabel {
                id
                value
                color
            }
            externalReferences {
                edges {
                    node {
                        id
                        standard_id
                        entity_type
                        source_name
                        description
                        url
                        external_id
                    }
                }
            }
            revoked
            confidence
            created
            modified
            name
            description
            aliases
            first_seen
            last_seen
            objective
            """;

    public Campaign(OpenCTIApiClient client) {
        super(client);
    }

    @Override
    protected String getEntityType() {
        return "Campaign";
    }

    @Override
    protected String getEntityName() {
        return "campaign";
    }

    @Override
    protected String getEntityNamePlural() {
        return "campaigns";
    }

    @Override
    protected String getProperties() {
        return PROPERTIES;
    }

    @Override
    protected String getOrderingEnum() {
        return "CampaignsOrdering";
    }

    /**
     * Create a new Campaign.
     *
     * @param name the campaign name (required)
     * @param params additional parameters
     * @return the created campaign
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> create(String name, Object... params) {
        if (name == null || name.isBlank()) {
            log.error("Missing required parameter: name");
            return null;
        }
        
        log.info("Creating Campaign: {}", name);
        
        Map<String, Object> input = buildInput(params);
        input.put("name", name);
        
        String mutation = """
            mutation CampaignAdd($input: CampaignAddInput!) {
                campaignAdd(input: $input) {
                    id
                    standard_id
                    entity_type
                    parent_types
                }
            }
            """;
        
        Map<String, Object> result = client.query(mutation, Map.of("input", input));
        Map<String, Object> data = (Map<String, Object>) result.get("data");
        return client.processMultipleFields((Map<String, Object>) data.get("campaignAdd"));
    }
}

File: opencti-platform/opencti-graphql/tests/01-unit/database/stix-2-0-converter-fixtures/SDOs/threats/campaign.ts
import type { StoreEntity } from '../../../../../../src/types/store';

export const CAMPAIGN_INSTANCE = {
  id: '11111111-1111-4111-8111-111111111111',
  standard_id: 'campaign--11111111-1111-4111-8111-111111111111',
  entity_type: 'Campaign',
  name: 'Campaign Stix 2.0',
  description: 'campaign description',
  created: '2025-01-01T00:00:00.000Z',
  modified: '2025-01-03T00:00:00.000Z',
  confidence: 80,
  revoked: false,
  aliases: ['camp-1'],
  first_seen: '2025-01-01T00:00:00.000Z',
  last_seen: '2025-01-02T00:00:00.000Z',
  objective: 'credential theft',
  createdBy: { standard_id: 'identity--18fe5225-fee1-5627-ad3e-20c14435b024' },
  objectMarking: [{ standard_id: 'marking-definition--89484dde-e3d2-547f-a6c6-d14824429eb1' }],
  objectLabel: [{ value: 'covid-19' }],
  externalReferences: [
    {
      source_name: 'capec',
      description: 'spear phishing',
      external_id: 'CAPEC-163',
    },
  ],
} as unknown as StoreEntity;

export const EXPECTED_CAMPAIGN = {
  id: 'campaign--11111111-1111-4111-8111-111111111111',
  type: 'campaign',
  spec_version: '2.0',
  x_opencti_id: '11111111-1111-4111-8111-111111111111',
  x_opencti_type: 'Campaign',
  x_opencti_granted_refs: [],
  x_opencti_files: [],
  created: '2025-01-01T00:00:00.000Z',
  modified: '2025-01-03T00:00:00.000Z',
  revoked: false,
  confidence: 80,
  labels: [
    'covid-19',
  ],
  created_by_ref: 'identity--18fe5225-fee1-5627-ad3e-20c14435b024',
  object_marking_refs: [
    'marking-definition--89484dde-e3d2-547f-a6c6-d14824429eb1',
  ],
  external_references: [
    {
      source_name: 'capec',
      description: 'spear phishing',
      external_id: 'CAPEC-163',
    },
  ],
  name: 'Campaign Stix 2.0',
  description: 'campaign description',
  aliases: ['camp-1'],
  first_seen: '2025-01-01T00:00:00.000Z',
  last_seen: '2025-01-02T00:00:00.000Z',
  objective: 'credential theft',
};
